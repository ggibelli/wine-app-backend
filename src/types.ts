export enum Province {
  AG = 'AG',
  AL = 'AL',
  AN = 'AN',
  AO = 'AO',
  AP = 'AP',
  AQ = 'AQ',
  AR = 'AR',
  AT = 'AT',
  AV = 'AV',
  BA = 'BA',
  BG = 'BG',
  BI = 'BI',
  BL = 'BL',
  BN = 'BN',
  BO = 'BO',
  BR = 'BR',
  BS = 'BS',
  BT = 'BT',
  BZ = 'BZ',
  CA = 'CA',
  CB = 'CB',
  CE = 'CE',
  CH = 'CH',
  CL = 'CL',
  CN = 'CN',
  CO = 'CO',
  CR = 'CR',
  CS = 'CS',
  CT = 'CT',
  CZ = 'CZ',
  EN = 'EN',
  FC = 'FC',
  FE = 'FE',
  FG = 'FG',
  FI = 'FI',
  FM = 'FM',
  FR = 'FR',
  GE = 'GE',
  GO = 'GO',
  GR = 'GR',
  IM = 'IM',
  IS = 'IS',
  KR = 'KR',
  LC = 'LC',
  LE = 'LE',
  LI = 'LI',
  LO = 'LO',
  LT = 'LT',
  LU = 'LU',
  MB = 'MB',
  MC = 'MC',
  ME = 'ME',
  MI = 'MI',
  MN = 'MN',
  MO = 'MO',
  MS = 'MS',
  MT = 'MT',
  NA = 'NA',
  NO = 'NO',
  NU = 'NU',
  OR = 'OR',
  PA = 'PA',
  PC = 'PC',
  PD = 'PD',
  PE = 'PE',
  PG = 'PG',
  PI = 'PI',
  PN = 'PN',
  PO = 'PO',
  PR = 'PR',
  PT = 'PT',
  PU = 'PU',
  PV = 'PV',
  PZ = 'PZ',
  RA = 'RA',
  RC = 'RC',
  RE = 'RE',
  RG = 'RG',
  RI = 'RI',
  RM = 'RM',
  RN = 'RN',
  RO = 'RO',
  SA = 'SA',
  SI = 'SI',
  SO = 'SO',
  SP = 'SP',
  SR = 'SR',
  SS = 'SS',
  SU = 'SU',
  SV = 'SV',
  TA = 'TA',
  TE = 'TE',
  TN = 'TN',
  TO = 'TO',
  TP = 'TP',
  TR = 'TR',
  TS = 'TS',
  TV = 'TV',
  UD = 'UD',
  VA = 'VA',
  VB = 'VB',
  VC = 'VC',
  VE = 'VE',
  VI = 'VI',
  VR = 'VR',
  VT = 'VT',
  VV = 'VV',
}

export enum Regioni {
  ABRUZZO = 'ABRUZZO',
  BASILICATA = 'BASILICATA',
  CALABRIA = 'CALABRIA',
  CAMPANIA = 'CAMPANIA',
  EMILIA = 'EMILIA',
  FRIULI = 'FRIULI',
  LAZIO = 'LAZIO',
  LIGURIA = 'LIGURIA',
  LOMBARDIA = 'LOMBARDIA',
  MARCHE = 'MARCHE',
  MOLISE = 'MOLISE',
  PIEMONTE = 'PIEMONTE',
  PUGLIA = 'PUGLIA',
  SARDEGNA = 'SARDEGNA',
  SICILIA = 'SICILIA',
  TOSCANA = 'TOSCANA',
  TRENTINO = 'TRENTINO',
  TRENTO = 'TRENTO',
  UMBRIA = 'UMBRIA',
  VALDAOSTA = 'VALDAOSTA',
  VENETO = 'VENETO',
}

export interface Address {
  via: string;
  CAP: string;
  comune: string;
  provincia: Province;
  regione: Regioni;
}

export enum TypeAd {
  SELL = 'SELL',
  BUY = 'BUY',
}

export enum MetodoProduttivo {
  CONVENZIONALE = 'CONVENZIONALE',
  BIOLOGICO = 'BIOLOGICO',
  BIODINAMICO = 'BIODINAMICO',
  NATURALE = 'NATURALE',
}

export enum TypeProduct {
  ADWINE = 'AdWine',
  ADGRAPE = 'AdGrape',
}

export enum EspressioneComunitaria {
  DOP = 'DOP',
  IGP = 'IGP',
  ND = 'ND',
}

export enum DenomZona {
  DOC = 'DOC',
  DOCG = 'DOCG',
  IGT = 'IGT',
  VARIETALE = 'VARIETALE',
  VINO = 'VINO',
}

export enum Menzione {
  CLASSICO = 'CLASSICO',
  RISERVA = 'RISERVA',
  SUPERIORE = 'SUPERIORE',
  VIGNA = 'VIGNA',
}

export enum QueryOrderBy {
  createdAt_ASC = 'createdAt_ASC',
  createdAt_DESC = 'createdAt_DESC',
  price_ASC = 'price_ASC',
  price_DESC = 'price_DESC',
}

export interface Errors {
  text: string;
  name: string;
}

export enum Rating {
  POOR = 'POOR',
  AVERAGE = 'AVERAGE',
  OK = 'OK',
  GOOD = 'GOOD',
  PERFECT = 'PERFECT',
}

export enum Colore {
  BIANCA = 'BIANCA',
  ROSSA = 'ROSSA',
}
