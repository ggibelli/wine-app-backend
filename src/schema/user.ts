import { gql } from 'apollo-server-express';

export const typeDefs = gql`
  type ProducedWines {
    wine: Wine!
    bottlesProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  type OwnedVineyards {
    vineyard: Vineyard!
    tonsProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input ProducedWinesInput {
    wine: ID!
    bottlesProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input OwnedVineyardsInput {
    vineyard: ID!
    tonsProduced: Int
    metodoProduttivo: MetodoProduttivo
  }

  input UserInput {
    email: String!
    password: String!
    firstName: String!
    lastName: String!
    pIva: String!
    phoneNumber: String!
    address: AddressInput!
    #verified: Boolean
    #premium: Boolean
    #ads: [ID!]
    #negotiations: [ID!]
    #reviews: [ID!]
    #adsRemaining: Int
    producedWines: ProducedWinesInput
    ownedVineyards: OwnedVineyardsInput
  }

  input UserInputUpdate {
    _id: ID!
    email: String
    firstName: String
    lastName: String
    pIva: String
    phoneNumber: String
    address: AddressInput
    isVerified: Boolean
    isPremium: Boolean
    ads: [ID!]
    negotiations: [ID!]
    reviews: [ID!]
    adsRemaining: Int
    producedWines: ProducedWinesInput
    ownedVineyards: OwnedVineyardsInput
  }

  type User {
    _id: ID!
    email: String!
    firstName: String!
    lastName: String!
    pIva: String!
    phoneNumber: String!
    address: Address!
    isVerified: Boolean! @authorized
    isPremium: Boolean @authorized
    isAdmin: Boolean!
    ads: [Ad!]
    negotiations: [Negotiation!]
    reviews: [Review!]
    adsRemaining: Int @authorized
    dateCreated: Date! @date
    producedWines: ProducedWines
    ownedVineyards: OwnedVineyards
  }

  type AuthUser {
    token: String!
    user: User!
  }

  type AuthUserPayload {
    response: AuthUser
    errors: [Error]
  }

  type UserPayload {
    response: User
    errors: [Error]
  }

  extend type Query {
    users: [User!] @authenticated
    user(id: ID!): User @authenticated
    me: User @authenticated
  }

  extend type Mutation {
    createUser(user: UserInput!): AuthUserPayload
    updateUser(user: UserInputUpdate, id: ID!): UserPayload
    deleteUser(id: ID!): UserPayload @authenticated
    login(email: String!, password: String!): AuthUserPayload
  }
`;
