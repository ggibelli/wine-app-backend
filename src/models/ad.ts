/* eslint-disable @typescript-eslint/no-unsafe-member-access */
import mongoose, { Schema, Document } from 'mongoose';
// import { Address } from '../types';
import {
  // Province,
  // Regioni,
  METODOPRODUTTIVO,
} from '../utils/enumMongooseHelper';
import { IUserDoc } from './user';
import { IWineDoc } from './wine';
import { TypeAd, TypeProduct, Menzione } from '../types';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { Negotiation } from '../generated/graphql';

export interface IAd {
  postedBy: IUserDoc['_id'];
  wineName?: string;
  wine?: IWineDoc['_id'];
  sottoZona?: string;
  menzione?: Menzione;
  metodoProduttivo?: MetodoProduttivo;
  vineyardName?: string;
  vineyard?: IVineyardDoc['_id'];
  harvest: number;
  abv: number;
  priceFrom: number; //se vendita priceFrom e priceTo settati a stesso numero
  priceTo: number;
  litersFrom?: number; //se vendite litersFrom e litersTo settati a stesso numero
  litersTo?: number;
  kgFrom?: number;
  kgTo?: number;
  content?: string | null | undefined;
  // address: Address;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'annuncio, solo attive, solo graphql??
  viewedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  savedBy?: mongoose.Types.Array<IUserDoc['_id']>;
  typeAd: TypeAd;
  typeProduct: TypeProduct;
  needsFollowUp: boolean;
  isActive: boolean;
  datePosted: Date;
}

export interface AdGraphQl extends IAd {
  _id: mongoose.Types.ObjectId | string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const adSchemaFields: Record<keyof IAd, any> = {
  postedBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  typeAd: {
    type: String,
    enum: [TypeAd.SELL, TypeAd.BUY],
    required: true,
  },
  typeProduct: {
    type: String,
    enum: [TypeProduct.ADWINE, TypeProduct.ADGRAPE],
    required: true,
  },
  wineName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
    index: true,
  },
  wine: {
    type: Schema.Types.ObjectId,
    ref: 'Wine',
    // required() {
    //   return this.typeProduct === TypeProduct.ADWINE;
    // },
  },
  sottoZona: {
    type: String,
  },
  menzione: {
    type: String,
    enum: ['CLASSICO', 'RISERVA', 'SUPERIORE', 'VIGNA'],
  },
  metodoProduttivo: {
    type: String,
    enum: Object.values(METODOPRODUTTIVO),
    required: true,
    default: 'CONVENZIONALE',
  },
  vineyardName: {
    type: String,
    minlength: 3,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
    index: true,
  },
  vineyard: {
    type: Schema.Types.ObjectId,
    ref: 'Vineyard',
    // required() {
    //   return this.typeProduct === TypeProduct.ADGRAPE;
    // },
  },
  abv: Number,
  harvest: {
    type: Number,
    required: true,
  },
  priceFrom: {
    type: Number,
    required: true,
  },
  priceTo: {
    type: Number,
    required: true,
  },
  litersFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  litersTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADWINE;
    },
  },
  kgFrom: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  kgTo: {
    type: Number,
    required() {
      return this.typeProduct === TypeProduct.ADGRAPE;
    },
  },
  // address: {
  //   comune: {
  //     type: String,
  //     required: true,
  //   },
  //   provincia: {
  //     type: String,
  //     enum: Object.values(Province),
  //     required: true,
  //   },
  //   regione: {
  //     type: String,
  //     enum: Object.values(Regioni),
  //     required: true,
  //   },
  // },
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  viewedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  savedBy: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  content: {
    type: String,
  },
  isActive: {
    type: Boolean,
    default: true,
  },
  needsFollowUp: {
    type: Boolean,
    default: true,
  },
  datePosted: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

export interface IAdDoc extends IAd, Document {}

const adSchema = new Schema(adSchemaFields);

adSchema.index({ typeProduct: 1, typeAd: 1, wineName: 1, vineyardName: 1 });

export const Ad = mongoose.model<IAdDoc>('Ad', adSchema);
