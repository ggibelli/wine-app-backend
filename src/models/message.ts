import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { INegotiationDoc } from './negotiation';

export interface IMessage {
  negotiation: INegotiationDoc['_id'];
  content: string;
  sentBy: IUserDoc['_id'];
  sentTo: IUserDoc['_id'];
  dateSent: Date;
  isViewed: boolean;
}

export interface MessageGraphQl extends IMessage {
  _id: mongoose.Types.ObjectId;
}

export interface IMessageDoc extends IMessage, Document {}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const messageSchemaFields: Record<keyof IMessage, any> = {
  negotiation: {
    type: Schema.Types.ObjectId,
    ref: 'Negotiation',
    required: true,
  },
  content: {
    type: String,
    minlength: 1,
    required: true,
  },
  sentBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  sentTo: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  dateSent: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  isViewed: {
    type: Boolean,
    default: false,
  },
};

const messageSchema = new Schema(messageSchemaFields);

messageSchema.index({ negotiation: 1, sentTo: 1, sentFrom: 1 });

export const Message = mongoose.model<IMessageDoc>('Message', messageSchema);
