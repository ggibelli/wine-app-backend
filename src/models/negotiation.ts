/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { IAdDoc } from './ad';
import { IMessageDoc } from './message';
import { TypeAd } from '../types';

export interface INegotiation {
  createdBy: IUserDoc['_id'];
  ad: IAdDoc['_id'];
  forUserAd: IUserDoc['_id'] | IUserDoc;
  messages?: Array<IMessageDoc['_id'] | IMessageDoc>;
  isConcluded: boolean;
  dateCreated: Date;
  type: TypeAd;
}

export interface INegotiationDoc extends INegotiation, Document {}

const negotiationSchemaFields: Record<keyof INegotiation, any> = {
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  type: {
    type: String,
    enum: ['Sell', 'Buy'],
    required: true,
  },
  ad: {
    type: Schema.Types.ObjectId,
    ref: 'Ad',
    required: true,
  },
  forUserAd: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  isConcluded: {
    type: Boolean,
    default: false,
  },
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
};

const negotiationSchema = new Schema(negotiationSchemaFields);

negotiationSchema.index({ createdBy: 1, ad: 1 }, { unique: true });

export const Negotiation = mongoose.model<INegotiationDoc>(
  'Negotiation',
  negotiationSchema
);
