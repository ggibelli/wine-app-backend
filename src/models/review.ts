/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import mongoose, { Schema, Document } from 'mongoose';
import { IUserDoc } from './user';
import { INegotiationDoc } from './negotiation';
import { TypeAd } from '../types';

enum Rating {
  POOR = 1,
  AVERAGE = 2,
  OK = 3,
  GOOD = 4,
  PERFECT = 5,
}

export interface IReview {
  createdBy: IUserDoc['_id'] | IUserDoc;
  negotiation: INegotiationDoc['_id'] | INegotiationDoc;
  forUserAd: IUserDoc['_id'] | IUserDoc;
  rating: Rating;
  dateCreated: Date;
  content: string;
  type: TypeAd;
}

export interface IReviewDoc extends IReview, Document {}

const reviewSchemaFields: Record<keyof IReview, any> = {
  createdBy: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  negotiation: {
    type: Schema.Types.ObjectId,
    ref: 'Negotiation',
    required: true,
  },
  forUserAd: {
    type: Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  content: {
    type: String,
    minlength: 5,
  },
  rating: {
    type: Number,
    enum: [1, 2, 3, 4, 5],
    required: true,
  },
  type: {
    type: String,
    enum: ['SELL', 'BUY'],
  },
};

const reviewSchema = new Schema(reviewSchemaFields);

reviewSchema.index({ createdBy: 1, negotiation: 1 }, { unique: true });

export const Review = mongoose.model<IReviewDoc>('Review', reviewSchema);
