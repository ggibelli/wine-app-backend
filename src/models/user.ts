/* eslint-disable @typescript-eslint/no-unsafe-return */
import mongoose, { Schema, Document, HookNextFunction } from 'mongoose';
import mongooseUniqueValidator from 'mongoose-unique-validator';
import { Address } from '../types';
import { Province, Regioni } from '../utils/enumMongooseHelper';
import { IAdDoc } from './ad';
import { IWineDoc } from './wine';
import { IReviewDoc } from './review';
import { IVineyardDoc } from './vineyard';
import { MetodoProduttivo } from '../types';
import { METODOPRODUTTIVO } from '../utils/enumMongooseHelper';
import bcrypt from 'bcrypt';
import { IMessageDoc } from './message';
import { Negotiation } from '../generated/graphql';

const HASH_ROUNDS = 10;

interface ProducedWines {
  wine: IWineDoc['_id'] | IWineDoc;
  bottlesProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

interface OwnedVineyards {
  vineyard: IVineyardDoc['_id'] | IVineyardDoc;
  tonsProduced: number;
  metodoProduttivo: MetodoProduttivo;
}

export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface IUser {
  email: string;
  firstName: string;
  lastName: string;
  password: string;
  pIva: string;
  phoneNumber: string;
  address: Address;
  coordinates?: Coordinates;
  isVerified: boolean;
  isPremium?: boolean;
  isAdmin: boolean;
  hideContact: boolean;
  ads?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  savedAds?: mongoose.Types.Array<IAdDoc['_id'] | IAdDoc>; // annunci postati dall'utente
  messages?: mongoose.Types.Array<IMessageDoc['_id'] | IMessageDoc>;
  negotiations?: mongoose.Types.Array<Negotiation>; // trattative dell'utente, attive e non, concluse e non
  reviews?: mongoose.Types.Array<IReviewDoc['_id'] | IReviewDoc>; // recensioni fatte e ricevute dall'utente
  adsRemaining?: number;
  dateCreated: Date;
  producedWines?: mongoose.Types.Array<ProducedWines>;
  ownedVineyards?: mongoose.Types.Array<OwnedVineyards>;
}

export interface IUserDoc extends IUser, Document {
  validatePassword(password: string): Promise<boolean>;
}

export interface UserGraphQl extends IUser {
  validatePassword(password: string): Promise<boolean>;
  _id: mongoose.Types.ObjectId;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const userSchemaFields: Record<keyof IUser, any> = {
  email: {
    type: String,
    unique: true,
    minlength: 3,
    required: true,
  },
  phoneNumber: {
    type: String,
    unique: true,
    minlength: 10,
    required: true,
  },
  firstName: String,
  lastName: String,
  password: String,
  pIva: {
    type: String,
    unique: true,
    minlength: 11,
    required: true,
  },
  address: {
    via: {
      type: String,
      required: true,
      minlength: 5,
    },
    CAP: {
      type: String,
      required: true,
      minlength: 5,
    },
    comune: { type: String, required: true, minlength: 5 },
    provincia: {
      type: String,
      enum: Object.values(Province),
      required: true,
    },
    regione: {
      type: String,
      enum: Object.values(Regioni),
      required: true,
    },
  },
  coordinates: {
    latitude: Number,
    longitude: Number,
  },
  isVerified: {
    type: Boolean,
    required: true,
    default: false,
  },
  isPremium: {
    type: Boolean,
    default: false,
  },
  isAdmin: {
    type: Boolean,
    default: false,
  },
  hideContact: {
    type: Boolean,
    default: true,
  },
  ads: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  savedAds: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Ad',
    },
  ],
  messages: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Message',
    },
  ],
  negotiations: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Negotiation',
    },
  ],
  reviews: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Review',
    },
  ],
  adsRemaining: Number,
  dateCreated: {
    type: Date,
    // `Date.now()` returns the current unix timestamp as a number
    default: Date.now,
  },
  producedWines: [
    {
      wine: {
        type: Schema.Types.ObjectId,
        ref: 'Wine',
      },
      bottlesProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
  ownedVineyards: [
    {
      vineyard: {
        type: Schema.Types.ObjectId,
        ref: 'Vineyard',
      },
      tonsProduced: Number,
      metodoProduttivo: {
        type: String,
        enum: Object.values(METODOPRODUTTIVO),
      },
    },
  ],
};

const userSchema = new Schema(userSchemaFields);

userSchema.pre('save', async function (next: HookNextFunction) {
  // here we need to retype 'this' because by default it is
  // of type Document from which the 'IUser' interface is inheriting
  // but the Document does not know about our password property
  const thisObj = this as IUserDoc;

  if (!this.isModified('password')) {
    return next();
  }

  try {
    const salt = await bcrypt.genSalt(HASH_ROUNDS);
    thisObj.password = await bcrypt.hash(thisObj.password, salt);
    return next();
  } catch (e) {
    return next(e);
  }
});

userSchema.methods.validatePassword = async function (pass: string) {
  return bcrypt.compare(pass, this.password);
};

userSchema.plugin(mongooseUniqueValidator);

export const User = mongoose.model<IUserDoc>('User', userSchema);
