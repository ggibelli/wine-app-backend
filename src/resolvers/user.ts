import {
  AdsResult,
  NegotiationResult,
  Resolvers,
  ReviewResult,
  UserAdsArgs,
  UserInput,
  UserInputUpdate,
  UserNegotiationsArgs,
  UserReviewsArgs,
} from '../generated/graphql';
import { UserGraphQl } from '../models/user';
import Users, { AuthResponse, UserResponse } from '../data-sources/users';
import Ads from '../data-sources/ads';
import Messages from '../data-sources/messages';
import Negotiations from '../data-sources/negotiations';
import { ForbiddenError } from 'apollo-server-express';
import Reviews from '../data-sources/reviews';
import { IAdDoc } from '../models/ad';
import { MessageGraphQl } from '../models/message';

interface StringIndexSignatureInterface {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  [index: string]: any;
}

type StringIndexed<T> = T & StringIndexSignatureInterface;

interface MongoDataSource {
  users: Users;
  ads: Ads;
  messages: Messages;
  negotiations: Negotiations;
  reviews: Reviews;
  // wines: Wines;
}

export const resolver: StringIndexed<Resolvers> = {
  Query: {
    async users(
      _,
      __,
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<UserGraphQl[]> {
      return dataSources.users.getUsers();
    },
    async user(
      _,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<UserGraphQl | undefined | null> {
      return dataSources.users.getUser(id);
    },
    me(_, __, { user }: { user: UserGraphQl }): UserGraphQl {
      console.log('yo');
      return user;
    },
  },
  Mutation: {
    async createUser(
      _,
      { user }: { user: UserInput },
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<AuthResponse> {
      return dataSources.users.createUser(user);
    },
    async updateUser(
      _,
      { user }: { user: UserInputUpdate },
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<UserResponse> {
      return dataSources.users.updateUser(user);
    },
    async deleteUser(
      _,
      { id }: { id: string },
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<UserResponse> {
      return dataSources.users.deleteUser(id);
    },
    async login(
      _,
      { email, password }: { email: string; password: string },
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<AuthResponse> {
      return dataSources.users.login(email, password);
    },
  },
  User: {
    async ads(
      user: UserGraphQl,
      args: UserAdsArgs,
      { dataSources }: { dataSources: MongoDataSource }
    ): Promise<AdsResult> {
      return dataSources.ads.getAdsByUser(user._id, args);
    },
    async messages(
      root: UserGraphQl,
      _,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ): Promise<MessageGraphQl[]> {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.messages.getMessages();
      }
      throw new ForbiddenError('You can only see your own messages');
    },
    async negotiations(
      root: UserGraphQl,
      args: UserNegotiationsArgs,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ): Promise<NegotiationResult> {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.negotiations.getNegotiations(args);
      }
      throw new ForbiddenError('You can only see your own negotiations');
    },
    async savedAds(
      root: UserGraphQl,
      _,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ): Promise<IAdDoc[]> {
      if (root._id.toString() === user._id.toString()) {
        if (user.savedAds?.length) {
          return dataSources.ads.findManyByIds(user.savedAds) as Promise<
            IAdDoc[]
          >;
        }
        return [];
      }
      throw new ForbiddenError('You can only see your own saved ads');
    },
    async reviews(
      root: UserGraphQl,
      args: UserReviewsArgs,
      { dataSources, user }: { dataSources: MongoDataSource; user: UserGraphQl }
    ): Promise<ReviewResult> {
      if (root._id.toString() === user._id.toString()) {
        return dataSources.reviews.getReviews(args);
      }
      throw new ForbiddenError('You can only see your own negotiations');
    },
    email(root: UserGraphQl, _, { user }: { user: UserGraphQl }): string {
      if (root._id.toString() === user._id.toString()) {
        return root.email;
      }
      if (!root.hideContact) {
        return root.email;
      }
      return 'The user prefers to hide the email';
    },
    phoneNumber(root: UserGraphQl, _, { user }: { user: UserGraphQl }): string {
      if (root._id.toString() === user._id.toString()) {
        return root.phoneNumber;
      }
      if (!root.hideContact) {
        return root.phoneNumber;
      }
      return 'The user prefers to hide the phone number';
    },
    // aggiungo logica per mostrare solo negotiation e certi dati solo se sono i miei
  },
};
