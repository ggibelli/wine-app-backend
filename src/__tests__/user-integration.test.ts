//import { gql } from 'apollo-server-express';
import {
  testClient,
  connectToDb,
  dropTestDb,
  closeDbConnection,
} from '../tests/integrationSetup';
import { User } from '../models/user';
import { users } from '../tests/mocksTests';
import { UserInput } from '../generated/graphql';
import { Province, Regioni } from '../types';

const { query, mutate } = testClient;

const LOGIN_VALID = `
  mutation {
    login(email: "gio@prova.it", password: "giovanni") {
      response {
        token
      }
      errors {
        name
      }
    }
  }
`;

const LOGIN_INVALID = `
  mutation {
    login(email: "wrong", password: "giovanni") {
      response {
        token
      }
      errors {
        name
      }
    }
  }
`;

const CREATE_USER = `
  mutation CreateUser($auth: UserInput!) {
    createUser(user: $auth) {
      response {
        token
      }
      errors {
        name
      }
    }
  }
`;

const indirizzo = {
  via: 'asd asddasd',
  CAP: 12345,
  comune: 'aaaa',
  provincia: Province.AT,
  regione: Regioni.PIEMONTE,
};

beforeAll(async () => {
  await connectToDb();
  await dropTestDb();
  const usersMock = users();
  const user = new User(usersMock[0]);
  await user.save();
});

describe('Integration test users', () => {
  it('valid login mutation returns token', async () => {
    const res = await query(LOGIN_VALID);
    expect(res).toMatchSnapshot();
    //expect(res.data.login.response.token).toBeDefined();
  });

  it('invalid login mutation returns errors', async () => {
    const res = await query(LOGIN_INVALID);
    expect(res).toMatchSnapshot();
    //expect(res.data.login.response).toBeNull();
    //expect(res.data.login.errors).toBeDefined();
  });

  it('create user mutation successfull', async () => {
    const auth: UserInput = {
      email: 'giollo@ollo.it',
      password: 'Pro@3###4~~vaPassWord!23',
      firstName: 'Giova',
      lastName: 'Gibelli',
      pIva: '01640250054',
      phoneNumber: '1234527890',
      address: indirizzo,
    };
    const res = await mutate(CREATE_USER, {
      variables: { auth: auth },
    });
    expect(res).toMatchSnapshot();
    //expect(res.data.createUser.response.token).toBeDefined();
    //expect(res.errors).toBeUndefined();
  });

  it('create user mutation returns errors', async () => {
    const auth: UserInput = {
      email: 'sbagliata#oo.it',
      password: 'facile',
      firstName: 'Giova',
      lastName: 'Gibelli',
      pIva: '0101010101',
      phoneNumber: '123',
      address: indirizzo,
    };
    const res = await mutate(CREATE_USER, {
      variables: { auth: auth },
    });
    expect(res).toMatchSnapshot();
    //expect(res.data.createUser.response).toBeNull();
    //expect(res.data.createUser.errors).toHaveLength(4);
  });
});

afterAll(async () => {
  await dropTestDb();
  await closeDbConnection();
});
