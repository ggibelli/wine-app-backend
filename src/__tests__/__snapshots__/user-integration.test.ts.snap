// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration test users create user mutation returns errors 1`] = `
Object {
  "data": Object {
    "createUser": Object {
      "errors": Array [
        Object {
          "name": "UserInputError",
        },
        Object {
          "name": "UserInputError",
        },
        Object {
          "name": "UserInputError",
        },
        Object {
          "name": "General Error",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test users create user mutation successfull 1`] = `
Object {
  "data": Object {
    "createUser": Object {
      "errors": Array [],
      "response": Object {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI1ZmJlZWNiYTJjNThiMzA1ODdhNTRmMzYiLCJpYXQiOjE2MDYzNDc5NjJ9._VxFHfjm4xpXhNRXzSj8qym4coqKkywGiuSBVvhJeJE",
      },
    },
  },
}
`;

exports[`Integration test users invalid login mutation returns errors 1`] = `
Object {
  "data": Object {
    "login": Object {
      "errors": Array [
        Object {
          "name": "UserInputError",
        },
      ],
      "response": null,
    },
  },
}
`;

exports[`Integration test users valid login mutation returns token 1`] = `
Object {
  "data": Object {
    "login": Object {
      "errors": Array [],
      "response": Object {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI1ZmJlZWNiYTJjNThiMzA1ODdhNTRmMzUiLCJpYXQiOjE2MDYzNDc5NjJ9.b83U7K1FueGZUVNIR1YaVXPf1CCa7cyVUtZ3OazN-4o",
      },
    },
  },
}
`;
